### DISCUSSION TITLE
how to convert scanned pdf to searchable pdf

### DISCUSSION THREAD

**Discussion by Laxmi530** [pymupdf/PyMuPDF#2474] [Looking for help]
Hai,
Actually, I am trying to convert a scanned pdf or non-searchable pdf to searchable pdf by using pytesseract. I am able to do if the pdf is one page pdf when it becomes multiple page it is doing only the last page.
Can someone please help me is there any method available in PyMuPDF. Below is the sample code and I am using.
```
file = r"C:\Users\name\Downloads\Scanned_pdf.pdf"
file_pdf = open('Converted_searchable.pdf', 'ab')
text_data = []
with fitz.open(file) as pdf:
    for page in pdf:
        _, _, img = pdf_text_box_image(page)
        text_data.append(pytesseract.image_to_pdf_or_hocr(img))
file_pdf.write(bytearray(item for item in text_data))
file_pdf.close()
```

Thanking you advance.

**Comment by JorjMcKie** [comment]
There is more than one option:
1. You can use the imprtable version of [OCRmyPDF](https://pypi.org/project/ocrmypdf/)
2. You can use PyMuPDF's built-in Tesseract support as follows:
```python
import fitz

DPI = 150  # desired resolution

src = fitz.open("input.pdf")
doc = fitz.open()  # output PDF with text layer

for page in src:
    pix = page.get_pixmap(dpi=DPI)
    imgpdf = fitz.open("pdf", pix.pdfocr_tobytes())  # make 1-page temp PDF with text layer
    doc.insert_pdf(imgpdf)  # append page
    imgpdf.close()

doc.save("input-ocr.pdf")
```

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie for the response.
I tried your code, but I am getting error. Can you please guide me where the issue is.

```
RuntimeError                              Traceback (most recent call last)
(file:///D:/Working/PDF_Extractor.ipynb) Cell 55 in ()
      for page in src:
          pix = page.get_pixmap(dpi=DPI)
--->      imgpdf = fitz.open("pdf", pix.pdfocr_tobytes())  # make 1-page temp PDF with text layer
          doc.insert_pdf(imgpdf)  # append page
          imgpdf.close()

File [~\AppData\Roaming\Python\Python39\site-packages\fitz\fitz.py:7298](https://file+.vscode-resource.vscode-cdn.net/d%3A/Working/~/AppData/Roaming/Python/Python39/site-packages/fitz/fitz.py:7298), in Pixmap.pdfocr_tobytes(self, compress, language)
   7286 """Save pixmap as an OCR-ed PDF page.
   7287 
   7288 Args:
   (...)
   7295     Tesseract's language support data.
   7296 """
   7297 if not TESSDATA_PREFIX:
-> 7298     raise RuntimeError("No OCR support: TESSDATA_PREFIX not set")
   7299 EnsureOwnership(self)
   7300 from io import BytesIO

RuntimeError: No OCR support: TESSDATA_PREFIX not set
```

Thanking you in advance.

**Reply by JorjMcKie** [reply]
You must set an environment variable to the language support folder of your Tesseract installation, see the [here](https://pymupdf.readthedocs.io/en/latest/installation.html). Ususally this
* Windows: `set TESSDATA_PREFIX=C:\Program Files\Tesseract-OCR\tessdata`
* Unix systems: `export TESSDATA_PREFIX=/usr/share/tesseract-ocr/4.00/tessdata`

Either temporarily before starting your script, or permanently, dependeing on your OS.

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie for the help.
- In local okay we can set the path but how can we do in the server while the code is getting executed from the container.
- How can we temporarily set an environment variable before starting our script.
Please help me these two concerns.

Thanking you in advance.

**Reply by JorjMcKie** [reply]
There no solutions for this in the current version.
The next version however will introduce a new parameter "tessdata" in OCR-related methods. This parameter allows to directly specify the name of said folder, for instance `pix.pdfocr_tobytes(tessdata=r"C:\Program Files\Tesseract-OCR\tessdata")`.
This version will also have a new function `fitz.get_tessdata()`, which tries to find the install location of Tesseract-OCR and then returns the tessdata folder name.

**Reply by JorjMcKie** [reply]
We aim to publish that new version (1.22.4) this week.

**Reply by Laxmi530** [reply]
Thanks @JorjMcKie that will be a great work. Please let me know once you update the library.

Thanking you in advance.

**Reply by SkaarFacee** [reply]
Adding to this I wanted to know if we have scope to use a OCR other than tesseract say maybe easyOCR or paddleOCR ?

**Reply by JorjMcKie** [reply]
No - only Tesseract has an in-built API in (py-) MuPDF.

**Reply by SkaarFacee** [reply]
Hey @JorjMcKie 👋🏻, thanks for the reply. Are there any plans to make it possible to input an raw ocr response and then be able to continue the task ?

**Reply by JorjMcKie** [reply]
There are no such plans at the time being.
